runtime/sam/op/robot/robot.go

```diff
@@ -0,0 +1,193 @@
+package robot
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/brimdata/super"
+	"github.com/brimdata/super/compiler/optimizer/demand"
+	"github.com/brimdata/super/pkg/storage"
+	"github.com/brimdata/super/runtime"
+	"github.com/brimdata/super/runtime/exec"
+	"github.com/brimdata/super/runtime/sam/expr"
+	"github.com/brimdata/super/zbuf"
+)
+
+type Op struct {
+	parent  zbuf.Puller
+	rctx    *runtime.Context
+	env     *exec.Environment
+	expr    expr.Evaluator
+	filter  zbuf.Filter
+	format  string
+	demand  demand.Demand
+	batch   zbuf.Batch
+	off     int
+	src     zbuf.Puller
+	targets []super.Value
+}
+
+func New(rctx *runtime.Context, env *exec.Environment, parent zbuf.Puller, e expr.Evaluator, format string, f zbuf.Filter, d demand.Demand) *Op {
+	return &Op{
+		parent: parent,
+		rctx:   rctx,
+		env:    env,
+		expr:   e,
+		demand: d,
+		filter: f,
+		format: format,
+	}
+}
+
+func (o *Op) Pull(done bool) (zbuf.Batch, error) {
+	if done {
+		if o.batch != nil {
+			o.batch.Unref()
+		}
+		o.batch = nil
+		src := o.src
+		o.src = nil
+		if closer, ok := src.(io.Closer); ok {
+			if err := closer.Close(); err != nil {
+				return nil, err
+			}
+		}
```

Comment:
Call Pull(true) when discarding a zbuf.Puller. (Neither runtime.Environment.Open nor OpenHTTP returns a value implementing io.Closer.)

```suggestion
		if o.src != nil {
			if _, err := o.src.Pull(true); err != nil {
				return nil, err
			}
		}
```